scripts/diff-pkg-changes.py

```diff
@@ -0,0 +1,351 @@
+                pkg_dict['Source'],
+                pkg_dict['Description'],
+                pkg_dict['OE'],
+                pkg_dict.get('HomePage', ''),
+                pkg_dict['License'],
+                pkg_dict['Priority']
+            ))
+        except KeyError:
+            raise UnexpectedError(
+                'Missing key info in Packages file of feed \'%s\' in branch %s.' \
+                    % (feed_name, branch_name)
+            )
+
+    return pkgs
+
+# Create a final report from all the package versions
+def gen_report(
+    old_core_pkgs: list['Package'], new_core_pkgs: list['Package'],
+    old_extra_pkgs: list['Package'], new_extra_pkgs: list['Package'],
+    old_vers: str, new_vers: str
+) -> dict:
+    # Handle core diffs
+    core_pkg_changes: list[dict] = []
+    core_pkgs_added: list[dict[str, str]] = []
+    core_pkgs_removed: list[dict[str, str]] = []
+    for pkg in old_core_pkgs:
+        other_vers_of_pkg: list['Package'] = \
+            [ i for i, other_pkg in enumerate(new_core_pkgs) if other_pkg.name == pkg.name ]
+        if len(other_vers_of_pkg) == 0: # Check for package removal from version to version
+            core_pkgs_removed.append({
+                'name': pkg.name,
+                'version': pkg.version
+            })
+        else:
+            other_pkg: 'Package' = new_core_pkgs[other_vers_of_pkg[0]] # We know there's only one FYI
+            diff: dict = pkg.diff(other_pkg)
+            if diff != {}:
+                core_pkg_changes.append(diff)
+            new_core_pkgs.pop(other_vers_of_pkg[0])
+    for pkg in new_core_pkgs:
+        # We removed packages that were duplicates, so what's left is new packages
+        core_pkgs_added.append({
+            'name': pkg.name,
+            'version': pkg.version
+        })
+
+    # Do the same for extra
+    extra_pkg_changes: list[dict] = []
+    extra_pkgs_added: list[dict[str, str]] = []
+    extra_pkgs_removed: list[dict[str, str]] = []
+    for pkg in old_extra_pkgs:
+        other_vers_of_pkg: list['Package'] = \
+            [ i for i, other_pkg in enumerate(new_extra_pkgs) if other_pkg.name == pkg.name ]
+        if len(other_vers_of_pkg) == 0: # Check for package removal from version to version
+            extra_pkgs_removed.append({
+                'name': pkg.name,
+                'version': pkg.version
+            })
+        else:
+            other_pkg: 'Package' = new_extra_pkgs[other_vers_of_pkg[0]] # We know there's only one FYI
+            diff: dict = pkg.diff(other_pkg)
+            if diff != {}:
+                extra_pkg_changes.append(diff)
+            new_extra_pkgs.pop(other_vers_of_pkg[0])
+    for pkg in new_extra_pkgs:
+        # We removed packages that were duplicates, so what's left is new packages
+        extra_pkgs_added.append({
+            'name': pkg.name,
+            'version': pkg.version
+        })
+
+    return {
+        (old_vers + '->' + new_vers): {
```

Comment:
This structure also needs to encode the architecture of the feed. We only support `x64` in hardknott, for now. But that might well change in the future, and we'll need to be able to distinguish between architectures.

Reply:
I noticed this too and wondered if we wanted this for sumo, but I think for a first pass on this script we can make an assumption that we care only about x64. i.e. as long as the architecture doesn't need a big rewrite. 
If we don't care about sumo, for the foreseeable future, there's only x64 and I don't think we should spend too much time accommodating something else. @amstewart ?

Reply:
There's only one place affected by architecture afaik, and that's part of the feed URL used for getting the Packages file. It has a '/x64/' right now. I think that's the only thing that would need to change to adjust architecture

Reply:
I didn't think of the different 'tunes' - `core2-64`, `x64` and `all`. I don't remember how that is handled by this script, if they all end up in the same list or not.

Reply:
@dylan-matthew-turner is it correct that you're parsing the Packages file only under http://nickdanger.amer.corp.natinst.com/feeds/2022Q3/x64/main/all/
and not also
http://nickdanger.amer.corp.natinst.com/feeds/2022Q3/x64/main/core2-64/ , and
http://nickdanger.amer.corp.natinst.com/feeds/2022Q3/x64/main/x64/ ?

If that's the case, you need to parse them all.

Reply:
> I think for a first pass on this script we can make an assumption that we care only about x64. i.e. as long as the architecture doesn't need a big rewrite.

I think it should only take one line or so to include the arch in the JSON structure. I think that's worth the effort at this stage, so that we don't have to rev the JSON structure version in the future.